internal/controller/watch.go

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) 2025, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package controller
+
+import (
+	"context"
+	"errors"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// HierarchyFunc is a generic function type that takes a context, client, and object as input and
+// returns an object of type T and an error.
+// This is used for MakeHierarchyWatchHandler to define the function that will be called to get the target object.
+type HierarchyFunc[T any] func(ctx context.Context, c client.Client, obj client.Object) (T, error)
+
+// MakeHierarchyWatchHandler is a function that creates a watch handler for a specific hierarchy.
+func MakeHierarchyWatchHandler[From client.Object, To client.Object](
+	hierarchyFunc HierarchyFunc[To],
+	c client.Client,
+	loggerMsg string,
+) func(ctx context.Context, obj client.Object) []reconcile.Request {
+	return func(ctx context.Context, obj client.Object) []reconcile.Request {
+		logger := log.FromContext(ctx)
+		logger.Info(loggerMsg)
+
+		fromObj, ok := obj.(From)
+		if !ok {
+			return nil
+		}
+
+		toObj, err := hierarchyFunc(ctx, c, fromObj)
+		if err != nil {
+			if errors.Is(err, &HierarchyNotFoundError{}) {
+				logger.Error(err, "Hierarchy not found", "source", fromObj)
```

Comment:
@Mirage20 removing these logs can make it more hard to troubleshoot isn't it? For example, if I get an error here, how do I know the actual error? we can't return errors from this.